librz/type/path.c

```diff
@@ -36,74 +36,167 @@ RZ_API void rz_type_path_free(RZ_NULLABLE RzTypePath *tpath) {
 	free(tpath);
 }
 
+static RzType *path_walker_parse_bracket(const RzTypeDB *typedb, RzType *parent, const char *path, size_t *i, st64 *offset) {
+	size_t nd = 0;
+	RzType *typd = parent;
+	st64 cur_dim_off = 1; // product of all dimension sizes
+	while (typd->kind == RZ_TYPE_KIND_ARRAY) {
+		++nd;
+
+		if (typd->array.count != 0 && cur_dim_off <= ST64_MAX / typd->array.count) { // check for overflow
+			cur_dim_off *= typd->array.count;
+		} else {
+			RZ_LOG_ERROR("Too many error subscriptions\n");
+			*offset = -1;
+			return NULL;
+		}
+
+		typd = typd->array.type;
+	}
+	cur_dim_off *= rz_type_get_base_type(typedb, typd)->size; // elem size in bits
+
+	typd = parent;
+	for (size_t id = 0; id < nd; ++id) {
+		if (path[*i] != '[') {
+			RZ_LOG_ERROR("Expected '[' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		++*i;
+		size_t tok_beg = *i;
+		for (; isdigit(path[*i]); ++*i)
+			;
+
+		if (path[*i] != ']') {
+			RZ_LOG_ERROR("Expected ']' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		char *idx_str = rz_str_ndup(&path[tok_beg], *i - tok_beg);
+		size_t idx = rz_num_math(NULL, idx_str);
+		free(idx_str);
```

Comment:
Maybe you are able to just pass `path + tok_beg` to `rz_num_math` or `rz_num_get` (to save the allocation/free)? 
At least `atol` stops and returns the number, when it finds the first non-digit character. But of cause the `rz_num_X` API is not documented properly \*sigh*, so I have no idea if it behaves the same (please add the doxygen to them if you look at it).

Reply:
As @XVilka mentioned, no need to document the `rz_num` stuff.

Reply:
Probably it is not the best idea. 
* `path + tok_beg` is not 0-terminated. 
* using functions like `atol` on non-0-terminated strings does not look good (even if we know what char goes after digit sequence).

Reply:
It doesn't need to be terminated, it is guaranteed that `tok_beg <= *i` and `!is_digit(path[*i])`. At least in your current code.
So the checks already happened, alternatively add an assert before with `rz_return_val_if_fail()`. Allocations are pretty resource extensive and not using them, if there is a simpler solution, is better.

Reply:
Here is how I will save the allocation.

The `rz_num_math` function internally determines the base of passed number. So using it allows to specify such paths: `"Struct.arr[0x20]"` which is useful from user's perspective. And it also frees me from determining the base myself.